### ISSUE TITLE
Numerical Instability?

### ISSUE THREAD

**Issue by cortner** [lab-cosmo/sphericart#178]
**EDIT:** tests shown lower down this thread suggest that for **bounded** inputs (e.g. norm <= 1) the numerical instability is fairly benign, and there should be no immediate concern for most practical use cases. 

I tested (for the first time today) the Julia kernels in F32 precision. They match the F64 results very poorly for increasing L. For L <= 4 the issue seems to be minor but not to be dismissed entirely. This suggests the method may have a numerical instability.  @CheukHinHoJerry has confirmed this happens also in the Python interface and will post his results I hope. 

I have not had time to investigate the cause. (poles??) I'll continue to update this thread regularly as I find out more, in case somebody has ideas.

The following results were obtained using 

https://github.com/lab-cosmo/sphericart/blob/jl/f32/julia/test/test_f32.jl

```
[ Info: L = 4
norm(Z - Z_f32, Inf) = 9.673968952483847e-5
norm((Z - Z_f32) ./ (1 .+ abs.(Z)), Inf) = 1.6170220597846877e-5
[ Info: L = 6
norm(Z - Z_f32, Inf) = 0.0041583938345866045
norm((Z - Z_f32) ./ (1 .+ abs.(Z)), Inf) = 0.00020217312748329195
[ Info: L = 8
norm(Z - Z_f32, Inf) = 0.10429337802634109
norm((Z - Z_f32) ./ (1 .+ abs.(Z)), Inf) = 0.0010284089140649845
[ Info: L = 10
norm(Z - Z_f32, Inf) = 2.4269896796904504
norm((Z - Z_f32) ./ (1 .+ abs.(Z)), Inf) = 0.005041625140445443
[ Info: L = 12
norm(Z - Z_f32, Inf) = 113.3912731744349
norm((Z - Z_f32) ./ (1 .+ abs.(Z)), Inf) = 0.02231292813299802
```

**Comment by cortner** [comment]
Another test comparing F64 with F128
```
[ Info: L = 4
norm(Z - Z_f32, Inf) = 0.0001287454243765751
norm(Z - Z_quad, Inf) = 1.95881209689249693683868663451520837e-13
[ Info: L = 8
norm(Z - Z_f32, Inf) = 0.21114671271061525
norm(Z - Z_quad, Inf) = 1.68267416383462080717987983387154378e-10
[ Info: L = 12
norm(Z - Z_f32, Inf) = 192.71228145062923
norm(Z - Z_quad, Inf) = 1.57384146581647470872186066134631652e-07
[ Info: L = 16
norm(Z - Z_f32, Inf) = 100425.36366271973
norm(Z - Z_quad, Inf) = 1.01009926417180620119650790789730260e-04
[ Info: L = 20
norm(Z - Z_f32, Inf) = 3.3747824560546875e7
norm(Z - Z_quad, Inf) = 6.24077383652671232730160680649000859e-02
```

**Comment by CheukHinHoJerry** [comment]
Seeing similar trend, 

with `n_samples = 10`:
![Image](https://github.com/user-attachments/assets/f34f7f7c-22aa-4570-ad64-1e23b677e2b2)

with `n_samples = 100_000`:
![Image](https://github.com/user-attachments/assets/3652acac-99cf-4b0d-b381-6425171af359)

Script that I used for benchmarking in `jax`.
```python
import jax
import jax.numpy as jnp
import sphericart.jax
import matplotlib.pyplot as plt

def compute_sph(xyz, l_max, dtype):
    xyz = xyz.astype(dtype)
    return sphericart.jax.solid_harmonics(xyz, l_max)

def benchmark_accuracy(n_samples=10, l_max=3):
    key = jax.random.PRNGKey(0)
    xyz_f64 = 6 * jax.random.uniform(key, (n_samples, 3), dtype=jnp.float64)

    # Compute solid harmonics
    jax.config.update("jax_enable_x64", True)
    sph_f64 = compute_sph(xyz_f64, l_max, jnp.float64)
    jax.config.update("jax_enable_x64", False)
    sph_f32 = compute_sph(xyz_f64.astype(jnp.float32), l_max, jnp.float32)

    # Compute errors
    jax.config.update("jax_enable_x64", True)
    abs_error = jnp.abs(sph_f64 - sph_f32)
    rel_error = abs_error / (jnp.abs(sph_f64) + 1e-12)

    return {
        "l_max": l_max,
        "max_abs_error": float(jnp.max(abs_error)),
        "mean_abs_error": float(jnp.mean(abs_error)),
        "max_rel_error": float(jnp.max(rel_error)),
        "mean_rel_error": float(jnp.mean(rel_error)),
    }

if __name__ == "__main__":
    l_max_list = [2, 4, 6, 8, 10]
    results = [benchmark_accuracy(l_max=l) for l in l_max_list]

    max_abs_errors = [r["max_abs_error"] for r in results]
    mean_abs_errors = [r["mean_abs_error"] for r in results]
    max_rel_errors = [r["max_rel_error"] for r in results]
    mean_rel_errors = [r["mean_rel_error"] for r in results]

    fig, axs = plt.subplots(2, 2, figsize=(10, 8))
    axs = axs.flatten()

    axs[0].plot(l_max_list, max_abs_errors, marker='o')
    axs[0].set_title("Max Absolute Error")
    axs[0].set_xlabel("l_max")
    axs[0].set_ylabel("Error")
    axs[0].set_yscale("log")

    axs[1].plot(l_max_list, mean_abs_errors, marker='o')
    axs[1].set_title("Mean Absolute Error")
    axs[1].set_xlabel("l_max")
    axs[1].set_ylabel("Error")
    axs[1].set_yscale("log")

    axs[2].plot(l_max_list, max_rel_errors, marker='o')
    axs[2].set_title("Max Relative Error")
    axs[2].set_xlabel("l_max")
    axs[2].set_ylabel("Relative Error")
    axs[2].set_yscale("log")

    axs[3].plot(l_max_list, mean_rel_errors, marker='o')
    axs[3].set_title("Mean Relative Error")
    axs[3].set_xlabel("l_max")
    axs[3].set_ylabel("Relative Error")
    axs[3].set_yscale("log")

    plt.tight_layout()
    plt.show()
```

**Comment by cortner** [comment]
this suggets the problem might be Q?
```
[ Info: L = 4
norm(Z - Z_f32, Inf) = 0.00012791609771056756
norm(Z - Z_quad, Inf) = 2.09260889379930120993237988881961576e-13
norm(∇Z - ∇Z_f32, Inf) = 8.212025317737523e-5
norm(∇Z - ∇Z_quad, Inf) = 1.04226341144538963622539898644015852e-13
[ Info: L = 8
norm(Z - Z_f32, Inf) = 0.13512698629347142
norm(Z - Z_quad, Inf) = 2.24048797849417351740906866261666633e-10
norm(∇Z - ∇Z_f32, Inf) = 0.2130936253358458
norm(∇Z - ∇Z_quad, Inf) = 2.83024256984543487667410344326300677e-10
[ Info: L = 12
norm(Z - Z_f32, Inf) = 228.87054577469826
norm(Z - Z_quad, Inf) = 4.88380013582853495976542404156429277e-07
norm(∇Z - ∇Z_f32, Inf) = 522.4016358235298
norm(∇Z - ∇Z_quad, Inf) = 1.18676763768286772752783557266970764e-06
[ Info: L = 16
norm(Z - Z_f32, Inf) = 69460.48294067383
norm(Z - Z_quad, Inf) = 7.51035751775427716269526051214884673e-05
norm(∇Z - ∇Z_f32, Inf) = 238773.3696979056
norm(∇Z - ∇Z_quad, Inf) = 2.84668470141684806044836061171240937e-04
[ Info: L = 20
norm(Z - Z_f32, Inf) = 5.0203737107421875e7
norm(Z - Z_quad, Inf) = 5.14910515737815479528653381247060850e-02
norm(∇Z - ∇Z_f32, Inf) = 2.2805617339802936e8
norm(∇Z - ∇Z_quad, Inf) = 2.80082427095607801920405451985891369e-01
```

**Comment by cortner** [comment]
Furthermore Jerry's tests showing that the mean error is much smaller suggests it is concentrated, hence speculation about the poles.

**Comment by ceriottm** [comment]
So is this at all related to #179 ?

**Comment by cortner** [comment]
I noticed this when starting the GPU kernels. I then realized I had bugs that didn't allow me to run F32. That latter is #179. 

Honestly, I now regret posting this before I finished understanding it. Yesterday already I added a note at the top that there is probably nothing to worry about.

**Comment by cortner** [comment]
I now think that the loss in accuracy is simply L^2 * r^{2L}, which is exactly how the solid harmonics scale as r, L gets large. If you agree that this is of no concern, then we could even close this issue now.

**Comment by ceriottm** [comment]
No problema, I'm very happy there're people with proper numerical analysis creds hanging around this repo. Let's wait until you're content there's nothing to worry about ,other than the fact that 10^100 is a very big number, and then we can close it. Ciao!

**Comment by cortner** [comment]
Ran a few more tests that confirm L^2 * r^{2L} scaling for the solid harmonics and L^3 r^{2L} for gradients. It would be nice to get rid of this, but I think that's now mostly academic. So until the issue surfaces in a real scenario, I'll close this.
